rescript/replace_taxonomy.py

```diff
@@ -9,15 +9,52 @@
 from qiime2.plugin import MetadataColumn, Categorical
 import warnings
 import pandas as pd
+import re
+from qiime2.plugin import List
+
+
+def make_search_replace_dict(search_strings, replacement_strings):
+    ssl = len(search_strings)
+    rsl = len(replacement_strings)
+
+    if ssl == rsl:
+        sr_dict = dict(zip(search_strings, replacement_strings))
+        return sr_dict
+    else:
+        raise ValueError('Search list is not the same length as the '
+                         'replacement list! The search and replacement '
+                         'lists contain %i and %i items, respectively.'
+                         % (ssl, rsl))
 
 
 def replace_taxonomy(taxonomy: pd.Series,
-                     replacement_map: MetadataColumn[Categorical],
-                     regex: bool = True,
+                     replacement_map: MetadataColumn[Categorical] = None,
+                     search_strings: List = None,
+                     replacement_strings: List = None,
+                     use_regex: bool = False,
                      ) -> pd.Series:
 
-    updated_tax_series = taxonomy.replace(replacement_map.to_series(
-                                                    ).to_dict(), regex=regex)
+    if search_strings and replacement_strings:
+        print('Processing strings from command line.')
+        rm_dict = make_search_replace_dict(search_strings, replacement_strings)
+    elif replacement_map:
+        print('Processing strings from taxonomy map file.')
+        rm_dict = replacement_map.to_series().to_dict()
+    else:
+        raise ValueError('Either a replacement-map or both search-strings '
+                         'and replacement-stings must be supplied.')
+
+    # `use_regex` will basically toggle between using, or not, `re.escape`
+    # prior to passing to `re.compile`. This will allow the user to
+    # search and replace substrings w/o having to be knowledgable about
+    # regular expresions. That is, passing through `re.escape` will behave as
+    # a literal string match.
+    new_rm_dict = {(re.compile(k) if use_regex else re.compile(
+                            re.escape(k))): v for k, v in rm_dict.items()}
+
+    # Note, becuase we are compiling a regex either with or without
+    # `re.escape`, we must always set `regex=True` for `pd.replace`.
+    updated_tax_series = taxonomy.replace(new_rm_dict, regex=True)
 
     # User sanity check. Though only appears with `--verbose`
     ranks_in = taxonomy.str.count(';')
```

Comment:
`ranks_in.compare(ranks_out)` will return a Series of the diff between `ranks_in` and `ranks_out`, with an index corresponding to the row number... so instead of merely warning that the number of ranks differ, we could actually specify which lines differ.

I am not saying that this needs to be done or would even be useful... just throwing the idea out there in case you think it is useful.

Reply:
I like this and implemented it into the warning. Trying to figure out how to test that the warning is raised.

Reply:
you can test the the warning is raised with [assertWarnsRegex](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertWarnsRegex)

here's an example from [q2-sample-classifier](https://github.com/qiime2/q2-sample-classifier/blob/a787ed020e7fc5e4bd0b045b563025d6bec9cf59/q2_sample_classifier/tests/test_estimators.py#L257)